### PR TITLE
fix(material/datepicker): avoid rerender when min/maxDate changes to different time on the same day

### LINKED ISSUE #24435: bug(material/datepicker): unable to select date on inline-calendar when passing a `new Date()` to the [minDate] Input
### Is this a regression?

#24384 introduces this bug

### Description

When passing a `new Date()` to the [minDate] Input on mat calendar, selectedDateChange is not emitted when clicking on a date with the mouse.

### Reproduction

Steps to reproduce:
1. Open the [Stackblitz example](https://stackblitz.com/edit/components-issue-c79f7p)
2. Click on any date on the calendar

### Expected Behavior

Should display the selected date below the calendar and log the current date

### Actual Behavior

Nothing happens, the selected date does not change, and nothing is logged to the console.

### Environment

Macos 12.2.1 (21D62), Chrome 98.0.4758.102

#### Additional notes from bug reported
I would not recommend passing a `new Date()` every time to the `minDate` or `maxDate` inputs, even if it didn't cause this bug. I'll write up this bug for it anyways as something to consider.

### PR DESCRIPTION
**Pull request by zarend**
Avoid re-rendering <mat-calendar/> when the [minDate] or [maxDate]
Input change to a different time on the same day.

In `ngOnChanges`, do not call `init` if the previous and current value for
minDate/maxDate are on the same day. This makes #24384 a non-breaking
code change.

Fixes #24435

### CODE REVIEW

File: src/material/datepicker/calendar.ts [hunk 1]
```diff
@@ -255,19 +256,43 @@ export class MatCalendar<D> implements AfterContentInit, AfterViewChecked, OnDes
...
   /** The maximum selectable date. */
   @Input()
   get maxDate(): D | null {
     return this._maxDate;
   }
   set maxDate(value: D | null) {
-    this._maxDate = this._dateAdapter.getValidDateOrNull(this._dateAdapter.deserialize(value));
+    this._oldMaxDate = this._maxDate;
+
+    const validated = this._dateAdapter.getValidDateOrNull(this._dateAdapter.deserialize(value));
```
Comment (wagnermaciel) [inline src/material/datepicker/calendar.ts:283]: Is it possible to move this logic into a separate function? We seem to be repeating this in minDate's setter

File: src/material/datepicker/calendar.ts [hunk 2]
```diff
@@ -393,7 +418,19 @@ export class MatCalendar<D> implements AfterContentInit, AfterViewChecked, OnDes
   }
 
   ngOnChanges(changes: SimpleChanges) {
-    const change = changes['minDate'] || changes['maxDate'] || changes['dateFilter'];
+    const minDateChange: SimpleChange | undefined = !this._dateAdapter.sameDate(
+      this._minDate,
+      this._oldMinDate,
```
Comment (wagnermaciel) [inline src/material/datepicker/calendar.ts:423]: Do _oldMinDate and _oldMaxDate exist for the sole purpose of comparing them here? If so, SimpleChange has a reference to the previous value you can use - https://angular.io/api/core/SimpleChange

**Review by wagnermaciel — CHANGES_REQUESTED** [review]

File: src/material/datepicker/calendar.ts (hunk 2 above)
Comment (zarend) [reply src/material/datepicker/calendar.ts]: done ✔️

**Comment by zarend** [comment]
@wagnermaciel @crisbeto this PR is ready for your eyes again 👀 . I've addressed the PR comments and it is no longer a draft. This is the most readable solution I could come up with, curious to know what you think.

File: src/material/datepicker/calendar.ts [hunk 3]
```diff
@@ -393,7 +393,27 @@ export class MatCalendar<D> implements AfterContentInit, AfterViewChecked, OnDes
   }
 
   ngOnChanges(changes: SimpleChanges) {
-    const change = changes['minDate'] || changes['maxDate'] || changes['dateFilter'];
+    let change = undefined;
+
+    for (let key in changes) {
```
Comment (andrewseguin) [inline src/material/datepicker/calendar.ts:398]: The use of the for loop that breaks when it finds a change is a little hard to follow. Would prefer to see more straightforward logic here, e.g. something more like

```
const minDateChanged = !this._dateAdapter.sameDate(changes['minDate'].previousValue, changes['minDate'].currentValue) && !changes['minDate'].firstChange;
const minDateChanged = !this._dateAdapter.sameDate(changes['maxDate'].previousValue, changes['maxDate'].currentValue) && !changes['maxDate'].firstChange;
const dateFilterChanged = changes['dateFilter'] && !changes['dateFilter'].firstChange

if (minDateChanged || maxDateChanged || dateFilterChanged) {
  ...
}

```

Comment (zarend) [reply src/material/datepicker/calendar.ts]: Okay, let's not use the for-loop approach.

**Review by andrewseguin — APPROVED** [review]
